### PR TITLE
Report failure on exception handler

### PR DESCRIPTION
**Pull request by vietj**
please review : this feature allows Vertx uncaught exception to be reported as test failures

### CODE REVIEW

File: src/test/java/io/vertx/ext/unit/JUnitTest.java [hunk 1]
```diff
@@ -940,4 +943,101 @@ public void testStaticRunTestOnContext() throws Exception {
...
+  public static class DeployFailingVerticle {
+
+    @Before
+    public void before(TestContext context) {
+      Async async = context.async();
+      Vertx vertx = Vertx.vertx().exceptionHandler(context.exceptionHandler());
+      vertx.deployVerticle(VerticleFailStart.class.getName(), ar -> {
+        assertTrue(ar.succeeded());
+        async.complete();
+      });
```
Comment (cescoffier) [inline src/test/java/io/vertx/ext/unit/JUnitTest.java]: Vert.x instance not closed.

**Comment by cescoffier** [comment]
So, if I get it right you still need to register the exceptionHandler in your tests. I saw that it can be done in the `@Before` class. That means that the `TestContext` is now the same in the `@Before` and in the actual test method (it was not the case before). 

What about the _run on event loop_ rule that creates the vert.x instance for you if I remember correctly ?

File: src/test/java/io/vertx/ext/unit/JUnitTest.java [hunk 2]
```diff
@@ -940,4 +943,101 @@ public void testStaticRunTestOnContext() throws Exception {
...
+    static final AtomicInteger count = new AtomicInteger();
+
+    @Test
+    public void testMethod1(TestContext context) {
+      Async async = context.async();
+      Vertx vertx = Vertx.vertx().exceptionHandler(context.exceptionHandler());
+      vertx.runOnContext(v -> {
+        count.incrementAndGet();
+        fail();
+      });
```
Comment (cescoffier) [inline src/test/java/io/vertx/ext/unit/JUnitTest.java]: vert.x instance not closed.

**Comment by vietj** [comment]
@cescoffier yes now the same context is shared between before/test/after to allow this to work. So a failure can happen in the HttpServer request handler setup and the before phase and make the test fail

With the @RunTestOnContext, the user needs to register the exceptionHandler on the Vertx instance created by the rule like in this case

**Comment by cescoffier** [comment]
Is the @RunTestOnContext thing documented ? (I may have missed it).

**Comment by vietj** [comment]
@RunTestOnContext is documented in the junit section

**Comment by cescoffier** [comment]
+1

**Comment by cescoffier** [comment]
Just a feedback:

it allows using Harmcrest or AssertJ in callbacks if the `exceptionHandler` is positioned correctly. And that's pretty cool !

**Comment by vietj** [comment]
thanks, will merge it today

**Pull request merged**